src/src/PluginStructs/P104_data_struct.cpp

```diff
@@ -0,0 +1,2458 @@
+
+  d = node_time.day();
+  m = node_time.month();
+  y = year;
+  # ifdef P104_USE_DATETIME_OPTIONS
+
+  if (showYear) {
+    switch (dateFmt) {
+      case P104_DATE_FORMAT_US:
+        d = node_time.month();
+        m = node_time.day();
+        y = year;
+        break;
+      case P104_DATE_FORMAT_JP:
+        d = year;
+        m = node_time.month();
+        y = node_time.day();
+        break;
+    }
+  } else {
+    if ((dateFmt == P104_DATE_FORMAT_US) ||
+        (dateFmt == P104_DATE_FORMAT_JP)) {
+      std::swap(d, m);
+    }
+  }
+  # endif // ifdef P104_USE_DATETIME_OPTIONS
+
+  if (showYear) {
+    sprintf_P(psz, PSTR("%02d%c%02d%c%02d"), d, sep, m, sep, y);     // %02d will expand to 04 when needed
+  } else {
+    sprintf_P(psz, PSTR("%02d%c%02d"), d, sep, m);
+  }
+}
+
+uint8_t getDateTime(char           *psz,
+                    bool            colon    = true,
+                    bool            time12h  = false,
+                    bool            timeAmpm = false,
+                    bool            fourDgt  = false
+                    # ifdef         P104_USE_DATETIME_OPTIONS
+                    , const uint8_t dateFmt = 0
+                    , const uint8_t dateSep = 0
+                    # endif // ifdef P104_USE_DATETIME_OPTIONS
+                    ) {
+  String   ampm;
+  uint16_t d, M, y;
+  uint8_t  h, m;
+  const uint16_t year = node_time.year() - (fourDgt ? 0 : 2000);
+
+  # ifdef P104_USE_DATETIME_OPTIONS
+  const String separators = F(" /-.");
+  const char   sep        = separators[dateSep];
+  # else // ifdef P104_USE_DATETIME_OPTIONS
+  const char sep = ' ';
+  # endif // ifdef P104_USE_DATETIME_OPTIONS
+
+  # ifdef P104_USE_DATETIME_OPTIONS
+
+  if (time12h) {
+    if (timeAmpm) {
+      ampm = (node_time.hour() >= 12 ? F("p") : F("a"));
+    }
+    h = node_time.hour() % 12;
+
+    if (h == 0) { h = 12; }
+  } else
+  # endif // ifdef P104_USE_DATETIME_OPTIONS
+  {
+    h = node_time.hour();
+  }
+  M = node_time.minute();
+
+  # ifdef P104_USE_DATETIME_OPTIONS
+
+  switch (dateFmt) {
+    case P104_DATE_FORMAT_US:
+      d = node_time.month();
+      m = node_time.day();
+      y = year;
+      break;
+    case P104_DATE_FORMAT_JP:
+      d = year;
+      m = node_time.month();
+      y = node_time.day();
+      break;
+    default:
+  # endif // ifdef P104_USE_DATETIME_OPTIONS
+  d = node_time.day();
+  m = node_time.month();
+  y = year;
+  # ifdef P104_USE_DATETIME_OPTIONS
+}
+
+  # endif // ifdef P104_USE_DATETIME_OPTIONS
+  sprintf_P(psz, PSTR("%02d%c%02d%c%02d %02d%c%02d%s"), d, sep, m, sep, y, h, (colon ? ':' : ' '), M, ampm.c_str()); // %02d will expand to
+                                                                                                                     // 04 when needed
+  return M;
+}
+
+# if defined(P104_USE_NUMERIC_DOUBLEHEIGHT_FONT) || defined(P104_USE_FULL_DOUBLEHEIGHT_FONT)
+void createHString(String& string) {
+  for (uint32_t i = 0; i < string.length(); i++) {
```

Comment:
Better declare the length as const variable first before the loop, or else you call `length()` on every loop.